content/blog/kube-dns-coredns-perfcomp.md

```diff
@@ -0,0 +1,114 @@
+
+Dnsmasq not caching negative responses may not seem like a big deal in a micro-services environment where processes are mostly querying for things that exist. But there is a quirk in Kubernetes pod DNS policy that results in a very high proportion of negative DNS responses.  The DNS policy causes a query for something like `google.com` to result in 3+ nonexistent name queries before actually trying `google.com` (e.g. `google.com.default.svc.cluster.local`, `google.com.svc.cluster.local`, `google.com.cluster.local`, `google.com.mylocal.com`). So in kube-dns, a query for `google.com` results in a series of queries of which only about 20% are cached, the other 80% have to make a comparatively expensive round trip to the kube-dns process.  In CoreDNS, all of the responses can be cached.  This can make a big difference in a cluster where many pods are communicating with one service.
+
+## Test Environment
+
+In a nutshell, I performed these tests using Kubernetes' DNS perf-test tool on a three node bare metal cluster using a private upstream DNS server. In more detail...
+
+**Authoritative Base Kube-dns and CoreDNS Manifests**:  I have based the configurations of kube-dns and CoreDNS on their [authoritative base templates](https://github.com/kubernetes/kubernetes/tree/master/cluster/addons/dns).  Specifically, tests use the authoritative default configurations for kube-dns (dnsmasq command flags in the Deployment) and CoreDNS (Corefile in the Configmap). A single instance of each DNS server was deployed.
+
+**Kubernetes Cluster**: The tests were performed in a three node kubernetes cluster, using the Calico network plugin.  The Nodes were hosted at bare metal provider [Packet](https://www.packet.net), using a Type 1 as the master, and Type 2 for the nodes.
+
+**Kubernetes DNS Perf-test Tool**: I used the [kubernetes dns perf-tests tool](https://github.com/kubernetes/perf-tests/tree/master/dns) modified to [allow tests on CoreDNS](https://github.com/kubernetes/perf-tests/pull/114).  The tool by default runs the DNS server under test on the master node and the client pod on separate node (node 1).  The tests are not performed against the cluster's active cluster DNS, instead they configure, spin up and tear down an instance for each set of tests.
+
+**Upstream Server**: The upstream server was a CoreDNS instance using the `template` plugin, configured to reply to all queries with answer of '1.2.3.4'.  It was run in the cluster, on a separate node from the client and DNS server under test (node 2).  This local server responds much more quickly than a real world upstream server would.
+
+**DNS Server Versions**: CoreDNS 1.1.3, kube-dns 1.10
```

Comment:
kube-dns 1.14.10?

Reply:
ah - ok.   I mistakenly thought it was on the same version track as kubernetes itself.